components/indexer.py

```diff
@@ -1,56 +1,137 @@
+            self.last_colour = Indexer.CargoColour.RED
+        elif self.has_blue():
+            self.last_colour = Indexer.CargoColour.BLUE
+
+        # Default state is for nothing to be moving and for the cat flap to be down
+        self._tunnel_direction = self._chimney_direction = Indexer.Direction.OFF
+        self._cat_flap_is_open = False
 
     @feedback
-    def has_back(self) -> bool:
-        return not self.prox_sensor1.get() or not self.prox_sensor2.get()
+    def has_cargo_in_chimney(self) -> bool:
+        return (
+            not self.lower_chimney_prox_sensor.get()
+            or not self.upper_chimney_prox_sensor.get()
+        )
 
     @feedback
-    def has_front(self) -> bool:
-        return self.colour_sensor.getProximity() > 400
+    def has_cargo_in_tunnel(self) -> bool:
+        return self.last_colour is not Indexer.CargoColour.NONE
 
     @feedback
-    def is_front_ours(self) -> bool:
-        raw = self.colour_sensor.getRawColor()
-        return (raw.red > raw.blue) == self.is_red
+    def are_we_red(self) -> bool:
+        return wpilib.DriverStation.getAlliance() == wpilib.DriverStation.Alliance.kRed
+
+    @feedback
+    def last_cargo_was_opposition(self) -> bool:
+        if self.last_colour is Indexer.CargoColour.RED and not self.are_we_red():
+            return True
+        elif self.last_colour is Indexer.CargoColour.BLUE and self.are_we_red():
+            return True
+
+        return False
+
+    @feedback
+    def ready_to_intake(self) -> bool:
+        # We cannot have a cargo in the tunnel, and we can't already have two cargo (one in chimney and one trapped)
+        if self.has_cargo_in_tunnel() or (
+            self.has_cargo_in_chimney() and self.has_trapped_cargo
+        ):
+            return False
+
+        return True
 
     @feedback
     def get_colours(self) -> str:
         raw = self.colour_sensor.getRawColor()
         return f"r{raw.red:.3f}, g{raw.green:.3f}, b{raw.blue:.3f}"
 
     @feedback
-    def get_speeds(self) -> Tuple[float, float]:
-        return self.speeds
+    def get_proximity(self) -> float:
+        return self.colour_sensor.getProximity()
+
+    @feedback
+    def get_last_colour(self) -> str:
+        return self.last_colour.name
+
+    def open_cat_flap(self) -> None:
+        self._cat_flap_is_open = True
+
+    def close_cat_flap(self) -> None:
+        self._cat_flap_is_open = False
+
+    def run_tunnel_motor(self, direction: Direction) -> None:
+        self._tunnel_direction = direction
+
+    def run_chimney_motor(self, direction: Direction) -> None:
+        self._chimney_direction = direction
 
-    def set(self, front: int, back: int) -> None:
-        self.speeds = (self.indexer_speed * front, self.indexer_speed * back)
+    def has_red(self) -> bool:
+        return self.colour_sensor.getRawColor().red > 900
 
-    def set_chute(self, on: bool) -> None:
-        # self.indexer_piston.set(on)
-        pass
+    def has_blue(self) -> bool:
```

Comment:
These names should refer to where the sensor is so it's obvious what the controller is doing, e.g. `tunnel_has_blue_ball`